renderdoc/driver/ihv/intel/intel_counters.cpp

```diff
@@ -0,0 +1,423 @@
+      }
+    }
+  }
+
+  return m_Counters;
+}
+
+uint32_t IntelCounters::GetPassCount()
+{
+  return (uint32_t)m_subscribedMetricSets.size();
+}
+
+uint32_t IntelCounters::BeginSession()
+{
+  if(!m_metricsDevice)
+    return 0;
+
+  m_passIndex = 0;
+  TMetricsDeviceParams_1_2 *deviceParams = m_metricsDevice->GetParams();
+  if(deviceParams->Version.MajorNumber < 1 ||
+     (deviceParams->Version.MajorNumber == 1 && deviceParams->Version.MinorNumber < 2))
+    return 0;
+
+  IOverride_1_2 *frequencyOverride = m_metricsDevice->GetOverrideByName("FrequencyOverride");
+  if(!frequencyOverride)
+    return 0;
+
+  TTypedValue_1_0 *maxFreqSymbol =
+      m_metricsDevice->GetGlobalSymbolValueByName("GpuMaxFrequencyMHz");
+  if(!maxFreqSymbol)
+    return 0;
+
+  TSetFrequencyOverrideParams_1_2 params;
+  params.Enable = true;
+  params.FrequencyMhz = maxFreqSymbol->ValueUInt32;
+  params.Pid = 0;
+  frequencyOverride->SetOverride(&params, sizeof(params));
+  Sleep(500);
+
+  return 0;
+}
+
+void IntelCounters::EndSesssion()
+{
+  if(!m_metricsDevice)
+    return;
+
+  TMetricsDeviceParams_1_2 *deviceParams = m_metricsDevice->GetParams();
+  if(deviceParams->Version.MajorNumber > 1 ||
+     (deviceParams->Version.MajorNumber == 1 && deviceParams->Version.MinorNumber >= 2))
+  {
+    IOverride_1_2 *frequencyOverride = m_metricsDevice->GetOverrideByName("FrequencyOverride");
+
+    if(frequencyOverride)
+    {
+      TSetFrequencyOverrideParams_1_2 params;
+      params.Enable = false;
+      params.Pid = 0;
+      frequencyOverride->SetOverride(&params, sizeof(params));
+    }
+  }
+}
+
+void IntelCounters::BeginPass()
+{
+  m_sampleIndex = 0;
+  TMetricSetParams_1_0 *metricSetParams = m_subscribedMetricSets[m_passIndex]->GetParams();
+  m_queryResult.resize(metricSetParams->MetricsCount + metricSetParams->InformationCount);
+}
+
+void IntelCounters::EndPass()
+{
+  m_passIndex++;
+}
+
+void IntelCounters::EnableCounter(GPUCounter index)
+{
+  uint32_t metricSetIdx = m_metricLocation[index].first;
+  IMetricSet_1_1 *metricSet = m_allMetricSets[metricSetIdx];
+  auto iter = std::find(m_subscribedMetricSets.begin(), m_subscribedMetricSets.end(), metricSet);
+  if(iter == m_subscribedMetricSets.end())
+  {
+    m_subscribedMetricSets.push_back(metricSet);
+    metricSetIdx = (uint32_t)m_subscribedMetricSets.size() - 1;
+  }
+  else
+  {
+    metricSetIdx = (uint32_t)(iter - m_subscribedMetricSets.begin());
+  }
+  m_subscribedMetricsByCounterSet[metricSetIdx].push_back(index);
+}
+
+void IntelCounters::DisableAllCounters()
+{
+  m_subscribedMetricSets.clear();
+  for(size_t i = 0; i < m_subscribedMetricsByCounterSet.size(); i++)
+    m_subscribedMetricsByCounterSet[i].clear();
+}
+
+void IntelCounters::BeginSample()
+{
+  if(!m_metricsDevice)
+    return;
+
+  D3D11_COUNTER_DESC counter_desc;
+  counter_desc.MiscFlags = 0;
+  counter_desc.Counter =
+      (D3D11_COUNTER)m_subscribedMetricSets[m_passIndex]->GetParams()->ApiSpecificId.D3D1XDevDependentId;
```

Comment:
So if I understand right, the main thing blocking this from working on at least GL would be calling the appropriate GL function and using the `ApiSpecificId.OGLQueryIntelId` / `ApiSpecificId.OGLQueryIntelName` to find the corresponding query, and I don't see a way to connect this up to Vulkan/D3D12. Is that along the right lines?

To be clear nothing that needs to change for this PR I just want to understand how this connects up.

Reply:
You are right about OpenGL, more details in INTEL_performance_query extension. Performance query API for Vulkan/D3D12 is not public yet. We have plans to support these APIs in future.

Reply:
OK great, I'll see if I can add that as it looks like a small extension.